### PR TITLE
feat: import additional data

### LINKED ISSUE #11: Import additional data
e.g. (omdb_data, wi)

this data is imported last in the final script and the import 

It is up to the developer to determine how to handle conflicting data (but it must be consistent)

### PR DESCRIPTION
**Pull request by MarLaslo**
#11 

Don't mind the DO, it should not be there.

Update works all the data is updated, all the empty fields and N/A are converted to null. 

Removed $ and commas from boxoffice so it is an integer now.

I have added poster attribute to the media table, as I saw in the source data paper that it is a requirement to have it. 

image

### CODE REVIEW

**Review by sofushn — CHANGES_REQUESTED** [review]
I like your use of the `CASE` statement 👍
Not sure if the PR is ready of review or not (guessing it is since its marked as open and not draft)

Here are couple of things that needs to be changed or added:
- `poster` - should be a promotional media of type `poster` (you can just set it for the main release)
- `rated` - the age rating of the movie (should be on the release of the movie, can see we forgot to add the column)
- `imdbrating` - Should be added as a new score with source set to 'IMDb'
- `language` - Is related to a release (there should be a table where )
- `released` - the imdb data only has release year, update the release matching the year with the right date or add new release if year doesn't match or there are multiple in the same year
- `production` - this looks like its the production company for a media
- `ratings` - create scores based on the values in the ratings column (looks like it already has a `source` and a `value`)
- `imdbvotes` - we should add an additional column to score for `vote_count`
- `metascore` - like with imdbrating (source = 'metacritic')

Can see you assume that `omdb_data` is in the `original` schema that I moved the data to, nice touch 👌

**Comment by MarLaslo** [comment]
Will fix the stuff!

**Comment by MarLaslo** [comment]
>     * `language` - Is related to a release (there should be a table where )

Language is in omdb is not a code but full words. I think we should decide if we want to have Language in code or in words. But I think it is waste of time to try to match the codes with language, if this is where you are pointing.  

I will skip this until we discuss.

**Comment by MarLaslo** [comment]
[DONE]
    
poster - should be a promotional media of type poster (you can just set it for the main release)
    rated - the age rating of the movie (should be on the release of the movie, can see we forgot to add the column)
    released - the imdb data only has release year, update the release matching the year with the right date or add new release if year doesn't match or there are multiple in the same year
    production - this looks like its the production company for a media

[TO BE DONE]
    imdbrating - Should be added as a new score with source set to 'IMDb'
    ratings - create scores based on the values in the ratings column (looks like it already has a source and a value)
    imdbvotes - we should add an additional column to score for vote_count
    metascore - like with imdbrating (source = 'metacritic')

[IGNORED]
    language - Is related to a release (there should be a table where )

**Comment by MarLaslo** [comment]
/*
***IDEA***
What if we create a table and have types in there?

related_media_category(categoryname, parentcategory)

We can have then images, websites, videos without parentcat
and subcategories like poster, actor, director, premiere for images
and etc... 
*/

**Review by idax6797 — CHANGES_REQUESTED** [review]
Looks fine, I guess when we have it all together, then we can test and see if every things connects the right way. 

Regarding:
- and for the promotional media, should we then make this a task for later?

**Comment by MarLaslo** [comment]
>> and for the promotional media, should we then make this a task for later?

If we agree tomorrow on it then yes, but I think this can be for the future of the portfolio.
This change would not require a lot of time 
Just creating a new table throwing in the types and parents
Then if we populate the db we will need to tweak the type value of import, only if we decide on different names as website and poster. 

Also thinking should we then remove the website from the media? as I assume it is a promotional media

**Comment by MarLaslo** [comment]
Table score is now updated with data from OMDB_DB 

Did not use metacatscore and imdbrating as I was using rating which had the same values plus rotten tomatoes.

**Comment by idax6797** [comment]
> > > and for the promotional media, should we then make this a task for later?
> 
> If we agree tomorrow on it then yes, but I think this can be for the future of the portfolio. This change would not require a lot of time Just creating a new table throwing in the types and parents Then if we populate the db we will need to tweak the type value of import, only if we decide on different names as website and poster.
> 
> Also thinking should we then remove the website from the media? as I assume it is a promotional media

Yes, I agree. I think the website should be removed from the media, as it makes more sense for it to be categorized under promotional media. If we decide to go with this.

**Review by idax6797 — APPROVED** [review]

File: src/omdb_data_merge.sql [hunk 1]
```diff
@@ -134,4 +99,116 @@ INSERT INTO media_production_company (media_id, production_company_id, type)
...
+/*
+***IDEA***
+What if we create a table and have types in there?
+
+related_media_category(categoryname, parentcategory)
+
+We can have then images, websites, videos without parentcat
+and subcategories like poster, actor, director, premiere for images
+and etc... 
+*/
```
Comment (sofushn) [inline src/omdb_data_merge.sql:214]: would generally just move this to the PR description or as a comment (personally preferer description as it is the first thing you see on a PR)

File: src/omdb_data_merge.sql [hunk 2]
```diff
@@ -134,4 +99,116 @@ INSERT INTO media_production_company (media_id, production_company_id, type)
...
+    UPDATE score s
+    SET 
+        value = j.value,
+        vote_count = j.imdbvotes
+    FROM json_extract j
+    WHERE j.media_id = s.media_id
+      AND j.source = 'IMDb'
+      AND s.vote_count < j.imdbvotes
+    RETURNING s.media_id, s.source, 'updated' AS status
+)
```
Comment (sofushn) [inline src/omdb_data_merge.sql:158]: Good idea!

File: src/omdb_data_merge.sql [hunk 3]
```diff
@@ -134,4 +99,116 @@ INSERT INTO media_production_company (media_id, production_company_id, type)
...
+                THEN REPLACE(o.imdbvotes, ',', '')::INTEGER
+            WHEN jsonb_array_length(o.ratings::jsonb) = 0
+                AND (o.imdbvotes != 'N/A' OR o.imdbvotes != '')
+                THEN REPLACE(o.imdbvotes,',','')::INTEGER
+            ELSE 0
+        END AS imdbvotes
+    FROM original.omdb_data o
+    JOIN media m ON m.imdb_id = o.tconst
+    LEFT JOIN LATERAL jsonb_array_elements(o.ratings::jsonb) AS json_data ON TRUE
+),
```
Comment (sofushn) [inline src/omdb_data_merge.sql:144]: Not sure I fully understand what is going on here but looks cool 😅

**Review by sofushn — APPROVED** [review]

File: src/omdb_data_merge.sql (hunk 1 above)
Comment (MarLaslo) [reply src/omdb_data_merge.sql:214]: Will change it I just added it to comments of PR

File: src/omdb_data_merge.sql (hunk 3 above)
Comment (MarLaslo) [reply src/omdb_data_merge.sql:144]: So i made this dummy table. For storing the data from rating column. 
As it is json array which had objects with value and source.
>> LEFT JOIN LATERAL jsonb_array_elements(o.ratings::jsonb) AS json_data ON TRUE
o.ratings::jsonb - converts string to json
jsonb_array_elements()splits the json array into separate objects

So the code creates separate rows for the elements in array. 
So I got the data for imdb, RT and MC in different rows of my dummy table. 

Then I used CASE to fix the data as it was in bad format. 
>>json_data->>'Source' 
This extracts the value from json object variable Source
and then just bunch of conditions and fixes as imdbvotes was string with comas between tousands
converting to integer after

And then I just used that table with corrected and extracted values for update and insert

**Pull request merged**